src/gg.regression.unity.bots/Runtime/Scripts/RGLegacyInputUtility/RGLegacyInputWrapper.cs

```diff
@@ -0,0 +1,251 @@
+using UnityEngine;
+
+namespace RegressionGames.RGLegacyInputUtility
+{
+    public static class RGLegacyInputWrapper
+    {
+        private static MonoBehaviour _simulationContext;
+        private static ISet<KeyCode> _keysHeld;
+        private static ISet<KeyCode> _newKeysDown;
+        private static ISet<KeyCode> _newKeysUp;
+        private static Dictionary<KeyCode, Coroutine> _removeNewCoroutines;
+        
+        /**
+         * If this flag is true (this is the initial state), then all inputs are
+         * forwarded to the regular UnityEngine.Input APIs.
+         * 
+         * If this flag is false (after calling StartSimulation), then the test driver
+         * has control of the inputs via the Simulate... methods and the game will
+         * no longer read inputs from the user's device.
+         */
+        public static bool IsPassthrough
+        {
+            get => _simulationContext == null;
+        }
+
+        /**
+         * Called by the test driver to take control of the user input.
+         * Context parameter is the driving MonoBehaviour to use as context.
+         */
+        public static void StartSimulation(MonoBehaviour context)
+        {
+            if (_simulationContext != null)
+            {
+                throw new InvalidOperationException("An active simulation is already in progress");
+            }
+
+            _simulationContext = context;
+            _keysHeld = new HashSet<KeyCode>();
+            _newKeysDown = new HashSet<KeyCode>();
+            _newKeysUp = new HashSet<KeyCode>();
+            _removeNewCoroutines = new Dictionary<KeyCode, Coroutine>();
+        }
+        
+        /**
+         * Called by the test driver to release control of the user input.
+         */
+        public static void StopSimulation()
+        {
+            if (_simulationContext != null)
+            {
+                foreach (Coroutine coro in _removeNewCoroutines.Values)
+                {
+                    _simulationContext.StopCoroutine(coro);
+                }
+
+                _newKeysDown = null;
+                _newKeysUp = null;
+                _keysHeld = null;
+                _removeNewCoroutines = null;
+                _simulationContext = null;
+            }
+        }
+        
+        /**
+         * Coroutine that clears "new" flag on a key (either down or up)
+         * on the next frame.
+         */
+        private static IEnumerator RemoveNewKey(KeyCode keyCode)
+        {
+            yield return null;
+            _newKeysDown.Remove(keyCode);
+            _newKeysUp.Remove(keyCode);
+        }
+
+        private static void ScheduleRemoveNewKey(KeyCode keyCode)
+        {
+            _removeNewCoroutines.Add(keyCode, _simulationContext.StartCoroutine(RemoveNewKey(keyCode)));
+        }
+        
+        private static void ClearKeyState(KeyCode keyCode)
+        {
+            if (_removeNewCoroutines.TryGetValue(keyCode, out Coroutine coro))
```

Comment:
we can replace `TryGetValue` with `Remove` to avoid accessing the dictionary twice here